### DISCUSSION TITLE
Wrapper script that needs paths to in-project executable

### DISCUSSION THREAD

**Discussion by LunarLambda** [mesonbuild/meson#10981] [Q&A] [answered]
My project has a wrapper script that uses objcopy to strip an executable to a raw binary, and then runs a locally-built 'fixup' program on it, meant for use with `custom_target`. Looks like this:

meson.build:
```meson
objcopy = find_program('objcopy')

gbafix = executable('gbafix', 'gbafix.c', native: true)

makerom = configure_file(
  input: 'makerom.in',
  output: 'makerom',
  configuration: {
    'OBJCOPY': objcopy.full_path(),
    'GBAFIX': gbafix.full_path(),
  })

exe = executable(...)

custom_target(
  'rom',
  input: exe,
  output: exe.name() + '.rom',
  command: [makerom, '@INPUT@', '@OUTPUT@'],
  build_by_default: true)
```

makerom.in:
```python
#!/usr/bin/env python3

OBJCOPY = '@OBJCOPY@'
GBAFIX = '@GBAFIX@'

import subprocess
from sys import argv

subprocess.call([OBJCOPY, '-O', 'binary'] + argv[1:3])
subprocess.call([GBAFIX] + argv[2:])
```

However, this can lead to build failures when the `custom_target` is built before `gbafix` is, depending on what order ninja compiles things in (multithreading issue).

Is there a way to solve this? I would find combining `meson.override_find_program()` and `find_program()` as a workaround quite hacky.

**Comment by eli-schwartz** [comment]
You could add gbafix as one of the inputs to `'rom'`, or list it in the `depends:` kwarg.

**Reply by LunarLambda** [reply]
Yeah, I considered this. I still think that's a somewhat "leaky" solution though, there's really no reason for the other code to interact with or even know about gbafix.

Is there anything that speaks against adding a `depends:` kwarg to `configure_file()`? It's fairly common for "post-build" custom_targets to use wrapper scripts that then need the paths to any involved executables.

**Reply by eli-schwartz** [reply]
A `depends:` of a configure_file would, if added, behave exactly like the existing `depfile:` kwarg and any files that the configured file depends on would cause `meson setup` to be rerun in order to recreate the configured file.

**Reply by LunarLambda** [reply]
Don't think that would help here, not sure. It's not about rebuilding things, but just building things in the right order to begin with. And the depfile is meant to be output by the script, anyhow. l

Maybe a dummy custom_target that depends on gbafix and the generated script could work? But I'm not sure if custom_targets are valid 'programs' by themselves, suitable for `command:` kwargs

**Reply by eli-schwartz** [reply]
I *know* it wouldn't help, that's why I described it the way I did. `depfile:` for `configure_file()` goes together with the `command:` kwarg, it's parsed at configure time as well and is incompatible with using `configuration_data()`. ;)

The point is that a configured file does not produce a build rule. At all. In any way. There's nothing to attach a dependency to.

> But I'm not sure if custom_targets are valid 'programs' by themselves, suitable for command: kwargs

Sure they are. :) It's in the documentation for `custom_target(command: ...)`.

Some people use this for generating scripts that are too complex to create via configure_file.

**Reply by xclaesse** [reply]
You could use custom_target() instead of configure_file(), that way you can add the depend, but you would have to write a small python script that does the replace in the file.

**Reply by LunarLambda** [reply]
yeah, seems like that would be the most straightforward way. thank you

**Reply by LunarLambda** [reply]
I came up with this now:

```meson
gbafix = executable(...)

_makerom = configure_file(..., output: 'makerom')

makerom = custom_target(
  'makerom',
  command: ['nop'],
  output: 'makerom',
  depends: [gbafix],
  build_by_default: true)

rom = custom_target(
  'rom'
  command: [makerom, '@INPUT@', '@OUTPUT@'],
  ...,
  build_by_default: true)
```

where `nop` is simply an empty python script. This way I have the dependency ordering of `custom_target`, and the convenience of `configure_file`. Since `configure_file` doesn't generate any build rules, all configured files have to exist before any targets compile, so this should... be okay? Assuming that `configure_file` doesn't check if the output is actually used..

**Reply by LunarLambda** [reply]
So, this does work, except it has the not-so-amusing quirk that now, `meson compile --clean` will actually delete the configured file, since meson believes it to be the output of the `custom_target`...

It'd be nice to have a clean way to achieve this pattern of "already existing/configured script file depends on other targets at runtime", I don't think it's terribly uncommon and losing out on `configure_file` *is* unfortunate.

It basically lies right in the middle between `custom_target` and `run_target`... It can be passed to a `command:`, and can create outputs (like `custom_target`, but doesn't have a dedicated build step (like `run_target`).

Maybe a dedicated `script_target`?

**Reply by eli-schwartz** [reply]
I don't see how that would help. This doesn't really have anything to do with meson. It has everything to do with the concept of a build outputs graph.

- "foo" is a dependency of "foobar"
- "bar" has no dependencies
- "bar" is a dependency of "baz"

etc.

In this case:
- "foo" is running meson setup
- "foobar" is a configure_file
- "bar" is the gbafix
- "baz" is the rom.

Your issue is that you want "baz" to depend on both "bar" and "foobar", but the latter is not part of the ***build graph***. You tried to fake it, but by adding it to the build graph, you caused ***ninja*** to... see it as part of the build graph, and thus consider it a candidate for pruning.

By the way, run_target is a part of the build graph. It's a node without children since it doesn't produce output. But it is a node.

What would a theoretical `script_target` do exactly?

**Reply by eli-schwartz** [reply]
I continue to believe the solution here is just adding "gbafix" as a depends of the rom target.

It's a parent node of the "rom", but it doesn't have any build graph relationship to "makerom". Even if the "makerom" was replaced by a custom target that output the same file using sed or something... You'd still be able to *build* "makerom" and "gbafix" in parallel, as the former only requires knowledge of the path that the latter will eventually be at.

**Reply by LunarLambda** [reply]
> What would a theoretical `script_target` do exactly?

When you pass a `custom_target` as a `command:`, the target's *output* is run.

This is an extra build step that's not needed, because the script is already "built" (either by already existing in the source tree, or as the result of being generated with `configure_file` or similar)

Instead all we care about is specifying the script file to run, and it's runtime dependencies. `run_target` is basically that, except that it cannot be used to create outputs, and therefore cannot be passed as a `command:`.

i.e. I want something like the following to "just work"

```meson
gbafix = executable(...)

script = configure_file(...)

# the makerom script already exists, but needs gbafix to run
# bikeshedded syntax
makerom = script_target('makerom', command: [script], depends: [gbafix])

# "downstream" build definitions should just be able to use makerom
# without needing to specify its dependencies

custom_target('rom', command: [makerom, '@INPUT@', '@OUTPUT@'], ...)
```

**Reply by eli-schwartz** [reply]
But run_target is literally nothing like that?

**Reply by LunarLambda** [reply]
In my mind it's similar insofar that the goal is to run the script without introducing another build step for it. Apologies if that was a confusing comparison.

**Reply by eli-schwartz** [reply]
But run_target() *does* add another build step for running the script.

The only difference between it and custom_target is that run_target's outputs are:
- a terminating node on the build graph, whose outputs permanently and irrevocably leave -- they are e.g. flashed to another device over a serial console, or copied to another machine via ssh, or they are something like clang-format that modifies source files.
- always considered stale / out of date, so rerunning the run_target will always, always, re-execute the command, and if you were allowed to depend on it, it would cause your makerom custom_target to be rebuilt on every single compilation attempt.


In a Makefile, this would be called a ".PHONY" target: https://www.gnu.org/software/make/manual/html_node/Phony-Targets.html

Internally, the "clean" target is a run_target, as is "test", "install", "uninstall", etc.

**Reply by LunarLambda** [reply]
Okay well, nevermind that then. Still, as said, I'd like something akin to the above snippet.

Guess I'll make my own subsitution script with `custom_target` and `env:`, or make the `custom_target` make a copy of the file.

**Comment by LunarLambda** [comment, marked as answer]
Ended up solving it like so:

`tools/copy`
```python3
#!/usr/bin/env python3

import shutil
from sys import argv

shutil.copy(argv[1], argv[2])
```

`tools/meson.build`
```meson
objcopy = find_program('objcopy')

gbafix = executable('gbafix', 'gbafix.c', native: true)

makerom = configure_file(
  input: 'makerom.in',
  output: 'makerom.in',
  configuration: {
    'OBJCOPY': objcopy.full_path(),
    'GBAFIX': gbafix.full_path(),
  })

makerom = custom_target('makerom',
  input: makerom,
  output: 'makerom',
  command: ['copy', '@INPUT@', '@OUTPUT@'],
  depends: [gbafix],
  build_by_default: true)
```

It's a hack, but it's the only reasonable way to let downstream projects use `makerom` without needing to worry about the implementation details.
